New lint: [`derive_partial_eq_without_eq`]

nsunderland1 opened the pull request:
Introduces a new lint, [`derive_partial_eq_without_eq`]. 

See: #1781 (doesn't close it though).

changelog: add lint [`derive_partial_eq_without_eq`]

rust-highfive commented:
r? @giraffate

(rust-highfive has picked a reviewer for you, use r? to override)

nsunderland1 commented on clippy_lints/src/derive.rs:
````diff
@@ -156,11 +156,44 @@ declare_clippy_lint! {
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub DERIVE_PARTIAL_EQ_WITHOUT_EQ,
+    nursery,
````
Wasn't sure about the right place to put this.

giraffate replied:
@rust-lang/clippy @dswij @Alexendoo @Jarcho Which category do you think is appropriate? I think `suspicious` looks good, but the original code isn't wrong.

nsunderland1 replied:
So this might be me being silly, but when I see a type that implements `PartialEq` but not `Eq`, I kind of read that as an assertion that "This is _not_ an equivalence relation, it's only a partial equivalence relation". In that sense I see it as suspicious to not implement `Eq` when you do have an equivalence relation, but that's pretty pedantic. I think you're correct in saying that this is primarily a question of adhering to API guidelines.

Jarcho replied:
The only reason not to implement `Eq` when it could be derived, other that laziness, is when there's the potential to extend the type in a way which could not implement `Eq`. This is not most cases, but it is a possibility. This possibility is, to me, enough of a reason to not categorize it as `suspicious`.

If not deriving `Eq` would cause the type to be misused, then `suspicious` would be a perfectly reasonable category. I think all our `suspicious` lints basically point out things which are easy to miswrite, or easy to misread in such a way that it creates a bug, but they are still reasonablish things to type. I haven't actually checked through all the `suspicious` lints, so that might not actually describe all `suspicious` lints.

flip1995 replied:
Completely agree with @Jarcho on this one. `sus` lints are for code that is "most likely wrong or useless". Not implementing `Eq` is not wrong, at least there is no way for Clippy to tell if it is intentional (possibly expanding the struct in the future). With that the API concern is the one left and for those, `style` is the best group.

giraffate replied:
Thanks for the comments!

nsunderland1 commented on clippy_lints/src/derive.rs:
```diff
@@ -419,3 +453,43 @@ impl<'tcx> Visitor<'tcx> for UnsafeVisitor<'_, 'tcx> {
+        then {
+            if !implements_trait(cx, ty, eq_trait_def_id, substs) {
+                // If all of our fields implement `Eq`, we can implement `Eq` too
+                for variant in adt.variants() {
```
This section is based off [can_type_implement_copy](https://github.com/rust-lang/rust/blob/37b55c8a0cafdb60b9168da34f904acc70157df8/compiler/rustc_trait_selection/src/traits/misc.rs#L19) from rustc. Only thing I wasn't super sure about is that that function does a [check](https://github.com/rust-lang/rust/blob/37b55c8a0cafdb60b9168da34f904acc70157df8/compiler/rustc_trait_selection/src/traits/misc.rs#L490) for `ty.references_error()`, it wasn't sure clear to me if that was necessary here (or indeed what exactly it does)

Jarcho replied:
I'm fairly sure you can ignore that. By the time late lint passes run the code already has no errors.

nsunderland1 commented on clippy_lints/src/derive.rs:
```diff
@@ -419,3 +453,43 @@ impl<'tcx> Visitor<'tcx> for UnsafeVisitor<'_, 'tcx> {
+                    }
+                }
+
+                span_lint_and_help(
```
I tried making this work with a suggestion like this:

```rust
span_lint_and_sugg(
    cx,
    DERIVE_PARTIAL_EQ_WITHOUT_EQ,
    span,
    "you are deriving `PartialEq` and can implement `Eq`",
    "consider deriving `Eq` as well",
    "PartialEq, Eq".to_string(),
    Applicability::MachineApplicable,
)
```

But when I ran this it just wouldn't print a suggestion at all, not sure what I did wrong.

Jarcho replied:
What did it print? That should have worked.

nsunderland1 replied:
So when I switch to `span_lint_and_sugg` on my branch, here's the stderr diff `cargo uitest` gives me: https://pastebin.com/v7Eg5kDp

Note that the only difference is that the ``help: consider deriving `Eq` as well`` is dropped.

Here's the JSON that `cargo uitest` spat out for one of the errors: https://pastebin.com/Gr6w967J

Looks like `suggested_replacement` and `suggestion_applicability` are null?

Jarcho replied:
It's because the span is for the impl block generated by the derive. You need to use `span.ctxt().outer_expn_data().call_site`.

nsunderland1 replied:
Ahhh, that makes a lot of sense, thanks! Fixed in the latest revision.

giraffate commented on clippy_lints/src/derive.rs:
```diff
@@ -176,6 +210,7 @@ impl<'tcx> LateLintPass<'tcx> for Derive {
 
             check_hash_peq(cx, item.span, trait_ref, ty, is_automatically_derived);
             check_ord_partial_ord(cx, item.span, trait_ref, ty, is_automatically_derived);
+            check_partial_eq_without_eq(cx, item.span, trait_ref, ty, is_automatically_derived);
```
Can this be moved to `if is_automatically_derived` block and `is_automatically_derived` be removed from args?

nsunderland1 replied:
Thanks for catching that, will fix

nsunderland1 commented:
Tentatively put the lint under `style` in the latest revision. As a side effect of this, I had to fix cases where tests were hitting the lint. In some of those I just added the derive, unless the tests cared specifically about whether `Eq` was implemented, in which case I just told them to ignore the lint. Tried to be fairly conservative there.

Also had to fix a few places in clippy where trivial enums were not being marked as `Eq`, and added a rustfix test for the lint.

nsunderland1 commented:
Took me a couple of tries to realize that `cargo test --features deny-warnings,internal` would catch everything for me :) Couple more places where this requires an `Eq` added in clippy internals, but all the cases are pretty straightforward.

giraffate commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit fe84ff336055412f0df8a2e625eaf46c6701e574 has been approved by `giraffate`

<!-- @bors r=giraffate fe84ff336055412f0df8a2e625eaf46c6701e574 -->
<!-- homu: {"type":"Approved","sha":"fe84ff336055412f0df8a2e625eaf46c6701e574","approver":"giraffate"} -->

bors commented:
:hourglass: Testing commit fe84ff336055412f0df8a2e625eaf46c6701e574 with merge d422baa30c415e6aed786f1953b0ac1b7d840dc1...
<!-- homu: {"type":"BuildStarted","head_sha":"fe84ff336055412f0df8a2e625eaf46c6701e574","merge_sha":"d422baa30c415e6aed786f1953b0ac1b7d840dc1"} -->
